keepalive.go

```diff
@@ -0,0 +1,125 @@
+package ssh
+
+import (
+	"log"
+	"sync"
+	"time"
+)
+
+type SessionKeepAlive struct {
+	clientAliveInterval time.Duration
+	clientAliveCountMax int
+
+	ticker       *time.Ticker
+	tickerCh     <-chan time.Time
+	lastReceived time.Time
+
+	metrics keepAliveMetrics
+
+	m      sync.Mutex
+	closed bool
+}
+
+type KeepAliveMetrics interface {
+	RequestHandlerCalled() int
+	KeepAliveReplyReceived() int
+	ServerRequestedKeepAlive() int
+}
+
+func NewSessionKeepAlive(clientAliveInterval time.Duration, clientAliveCountMax int) *SessionKeepAlive {
+	var t *time.Ticker
+	var tickerCh <-chan time.Time
+	if clientAliveInterval > 0 {
+		t = time.NewTicker(clientAliveInterval)
+		tickerCh = t.C
+	}
+
+	return &SessionKeepAlive{
+		clientAliveInterval: clientAliveInterval,
+		clientAliveCountMax: clientAliveCountMax,
+		ticker:              t,
+		tickerCh:            tickerCh,
+		lastReceived:        time.Now(),
+	}
+}
+
+func (ska *SessionKeepAlive) RequestHandlerCallback() {
+	ska.m.Lock()
+	ska.metrics.requestHandlerCalled++
+	ska.m.Unlock()
+
+	log.Println("ska.RequestHandlerCallback()")
+
+	ska.Reset()
+}
+
+func (ska *SessionKeepAlive) ServerRequestedKeepAliveCallback() {
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	// log.Println("ska.ServerRequestedKeepAliveCallback()")
+
+	ska.metrics.serverRequestedKeepAlive++
+}
+
+func (ska *SessionKeepAlive) Reset() {
+	ska.m.Lock()
+	defer ska.m.Unlock()
+
+	// log.Println("ska.Reset()")
+
+	ska.metrics.keepAliveReplyReceived++
+
+	if ska.ticker != nil && !ska.closed {
+		ska.lastReceived = time.Now()
+		ska.ticker.Reset(ska.clientAliveInterval)
```

Comment:
ensure that `ska.clientAliveInterval > 0` otherwise this will panic

Reply:
That condition is in the `NewSessionKeepAlive` constructor. Do you think that we need one more here?

Reply:
Missed that!